### ISSUE TITLE
Provide composer-rest-server support for passport-jwt strategy 

### ISSUE THREAD

**Issue by DennisM330** [hyperledger/composer#2038]
## Context
Provide documentation and support for the passport JWT strategy.  JWT is being used more and more in the API Economy and for protecting MicroServices in the following scenarios:

Authentication: This is the most common scenario for using JWT. Once the user is logged in, each subsequent request will include the JWT, allowing the user to access routes, services, and resources that are permitted with that token. Single Sign On is a feature that widely uses JWT nowadays, because of its small overhead and its ability to be easily used across different domains.

Information Exchange: JSON Web Tokens are a good way of securely transmitting information between parties. Because JWTs can be signed—for example, using public/private key pairs—you can be sure the senders are who they say they are. Additionally, as the signature is calculated using the header and the payload, you can also verify that the content hasn't been tampered with.

## Expected Behavior
The Hyperledger Composer Rest Server should validate the JWT token and, if valid, treat the session as authenticated so a HTTP 401 Authorization required error is not returned when trying to invoke an operation. 

## Actual Behavior
The Hyperledger Composer Rest Server returns a HTTP 401 Authorization required when trying to invoke an operation. 

## Possible Fix
Provide documentation and working example on the values needed for the COMPOSER_PROVIDERS environment variable . COMPOSER_PROVIDERS variable needs some of the JWT Strategy options as described in this link: (secretorkey, issuer, audience, altgorithms)
https://www.npmjs.com/package/passport-jwt

## Existing issues
This is an expansion of the passport strategy support referenced in the following issue:
https://github.com/hyperledger/composer/issues/142

## Context
Trying to take a generated JWT token from an API Gateway and pass the token to a secured composer-rest-server so that the composer-rest-server can validate the JWT token. 

## Your Environment
* Version used:
* Hyperledger Composer Rest Server 0.11.3
* Ubuntu Linux 16.04.2 LTS/Mac OSX 10.12.4
* API Gateway

**Comment by mrosack** [comment]
I've been able to get JWT authentication to work, adding some notes on the steps I used here...

- Since passport-jwt requires a function argument, a custom strategy is necessary.  Create a file called custom-jwt.js with the following contents: 
```
'use strict';

const passportJwt = require('passport-jwt');
const util = require('util');

function CustomJwtStrategy(options, verify) {
    options.jwtFromRequest = passportJwt.ExtractJwt.fromAuthHeaderAsBearerToken();
    passportJwt.Strategy.call(this, options, verify);
}
util.inherits(CustomJwtStrategy, passportJwt.Strategy);

module.exports = {
    Strategy: CustomJwtStrategy
};
```

- Use the following configuration in COMPOSER_PROVIDERS, making sure to use the correct relative path on your machine to custom-jwt.js.  The path is relative from `node_modules/loopback-component-passport/lib/passport-configurator.js`:

```
"jwt": {
  "provider": "jwt",
  "module": "../../../custom-jwt.js",
  "secretOrKey": "--your secret--",
  "authScheme": "saml",
  "successRedirect": "/",
  "failureRedirect": "/"
}
```

- Send a get request to /auth/jwt/callback, and the access_token cookie will be written to your browser, which can be used for future requests!

**Comment by DennisM330** [comment]
Will the access_token cookie be accessible in the headers for the HTTP Response? Assume the get request to /auth/jwt/callback containing the JWT token in the Authorization header is issued by an API Gateway.

**Comment by gangachris** [comment]
Found an interesting thread related to this on the Hyperledger Fabric Rocketchat 
https://chat.hyperledger.org/channel/composer?msg=KW4DbESMZKkPRWmPQ

**Comment by gangachris** [comment]
FYI, the jwt token passed in `Authorization Bearer <TOKEN>` needs to have a `username` field when it's being generated. I did not have this and kept getting a `Type Error: Cannot read property 'match' of undefined`

**Comment by DennisM330** [comment]
The context of this request is for the composer-rest-server to accept a token that is generated outside of the composer-rest-server by an API gateway which provides a rich set of features for creating and managing access tokens. The API gateway would generate the token, enforce how long the token is valid, revocation,etc. If the composer-rest-server determines the token is valid, the token is cached, and use to associate participant identities with the token.

**Comment by bhaveshopenxcell** [comment]
Hello @gangachris Thank you for your inputs. I am able to resolve the above module not found. Everything runs fine and not getting any issues even with "docker logs rest" command.

However, now when I am trying to access the "localhost:3000/auth/jwt" getting **Unauthorized** only. Can you please help me what could be the wrong thing here?

**Comment by gangachris** [comment]
@bhaveshopenxcell based on the above approach, you need to have a jwt token generated somewhere else with the same secret as the one passed in the config above. 

That way, when you call `localhost:3000/auth/jwt/callback`, with the jwt token passed, as either a header or a query params based on this line in your custom-jwt 
```
options.jwtFromRequest = passportJwt.ExtractJwt.fromAuthHeaderAsBearerToken();
```
 (in ☝️  it's in the header as Bearer Token, so you need to use a tool like postman, and pass in the header 'Authorization: Bearer jwt-token').

If you want to use the browser, change `fromAuthHeaderAsBearerToken` to `fromUrlQueryParameter("token")`, and then visit `http://localhost:3000/auth/jwt/callback?token=<token>`.

Regardless of the approach you take, the reponse is always a whole page, with cookies, and one of the cookies contains the new token you will use moving foreward for your requests. So you may need to find a way of extracting it. 

Hope this makes sense

**Comment by bhaveshopenxcell** [comment]
Hello @gangachris Thank you for your inputs. It really helped me to understand the flow. However, I do not get an idea when you say, 

>based on the above approach, you need to have a jwt token generated somewhere else with the same secret as the one passed in the config above.
I am wondering that is the only thing causing the issue. Can you please let me know how actually I should go from start accessing rest end points, once the above things has been set up? 
Do I need to create jwt token by any other code?

**Comment by saurabhgoel003** [comment]
Hello @gangachris. Thanks for providing the Dockerfile and docker env configuration. It is able to resolve the module- custom-jwt.js but getting following error on rest-server startup- 
````
TypeError: AuthStrategy is not a constructor
    at PassportConfigurator.configureProvider (/home/composer/.npm-global/lib/node_modules/composer-rest-server/node_modules/loopback-component-passport/lib/passport-configurator.js:420:18)
    at Promise.then (/home/composer/.npm-global/lib/node_modules/composer-rest-server/server/server.js:147:38)
    at <anonymous>
    at process._tickDomainCallback (internal/process/next_tick.js:228:7)
````
Following are the contents of custom-jwt.js module-

````
'use strict';

const passportJwt = require('passport-jwt');
const util = require('util');

function CustomJwtStrategy(options, verify) {
    options.jwtFromRequest = passportJwt.ExtractJwt.fromAuthHeaderAsBearerToken();
    passportJwt.Strategy.call(this, options, verify);
}
util.inherits(CustomJwtStrategy, passportJwt.Strategy);

module.exports = {
    Strategy: CustomJwtStrategy
};
````
Will much appreciate any pointers to what I might be doing wrong here. Thanks!

**Comment by gangachris** [comment]
I tried to cover this issue here https://www.codementor.io/gangachris125/passport-jwt-authentication-for-hyperledger-composer-rest-server-jqfgkoljn

**Comment by djraj** [comment]
Hello @gangachris I tried your method using Postman but unfortunately I only got back one cookie (connect.sid). I don't get back the access_token. Do you know what may be the issue?

**Comment by saurabhgoel003** [comment]
@djraj ... in response to /auth/jwt/callback, composer-rest-server sends a 302 redirect... that's why you are not able to catch all the cookies. Try making the get call with an additional property- maxRedirects:0.

    const authJwtHeader = {
        headers: {'authorization': 'Bearer '+<Your Token>},
        maxRedirects: 0,
        validateStatus: function(status) {
            return status >= 200 && status < 303;
        }
    };

    axios.get('http://localhost:3000/auth/jwt/callback', authJwtHeader).then(resp => {}, err => {});

validateStatus would be helpful when you will be making above call through some client and will want the 302 (Redirect response) catched as a success and not as an error.

**Comment by djraj** [comment]
@saurabhgoel003 Thank you for your reply. I am using a Java client. I have managed to dig through and found the error why the access token is not returning. As @gangachris mentioned, the composer-rest-server needs an "id" in the bearer token. My token generator generates a token using the id I provided but puts it as "jti" instead of "id". After manually changing them and testing using Postman, I managed to get the access token back. Thank you :)

**Comment by djraj** [comment]
Hello @saurabhgoel003 . I am able to get the access token via Postman. However, i still face the same issue of only "connect.sid" cookie and no access token returned using Java client. I have also set it not to follow redirects. Any ideas why?

**Comment by saurabhgoel003** [comment]
@djraj ... might be possible that the JWT authentication is not getting through on composer rest server. Did you observe any error / success logs on composer-rest-server related to authentication request??

**Comment by djraj** [comment]
Hello @saurabhgoel003 . I've noticed no unusual logs on composer-rest-server, just '401 - Unauthorized' error due to missing access token when a request is made. In another website [Codementor](https://www.codementor.io/gangachris125/passport-jwt-authentication-for-hyperledger-composer-rest-server-jqfgkoljn) , @gangachris said : 

> I also encountered a similar problem with a Go client. I had to use the built-in cookiejar package, instead of the normal cookie http cookie, because cookiejar persisits through redirects.
> -Christopher Ganga

I have tried retrieving the cookies using `CookieManager` method and also `getHeaders()`,`getCookies()` methods but I still could not grab all cookies. Perhaps there's any other cookie package for Java? Or any ideas how to get by this?

**Comment by derekm110** [comment]
@djraj I followed the tutorial from @gangachris, which was fantastic. I too, however, found myself only receiving the 'connect.sid' cookie. It was only after seeing @saurabhgoel003's comment about multiple redirects that I was able to achieve the desired behavior.

My implementation was slightly different, however, as I was using `request` rather than`axios`:

```
// the value of jwtSecretKey must match what is defined in your COMPOSER_PROVIDERS environment variable
let token = jwt.sign({ username: username, timestamp: timestamp }, jwtSecretKey)

let options = { 
    method: 'GET',
    url: 'http://[rest-server-ip]:3000/auth/jwt/callback',
    maxRedirects: '3', 
    followRedirect: false,
    headers: { 
        'Cache-Control': 'no-cache',
        Authorization: 'Bearer ' + token
    } 
};

request(options, function (error, response, body) {
    // upon a successful request, cookies are stored in response.headers['set-cookie']
})
```

They keys for me were setting `maxRedirects: 3` and `followRedirect: false`. 

At this point, I was able to retrieve all cookies from the `response.headers['set-cookie']` array (access_token, userId, and connect.sid)

I'm not a Java programmer, but does your HTTP client allow you to alter the way redirects are handled? If so, I would follow @djraj's advice and look further into that.

**Comment by johnjayasingh** [comment]
I have published a package called passport-empty for this fake jwt authentication instead of using

`"jwt": {
  "provider": "jwt",
  "module": "../../../custom-jwt.js",
  "secretOrKey": "--your secret--",
  "authScheme": "saml",
  "successRedirect": "/",
  "failureRedirect": "/"
}`

use this module to avoid storing/moving the module file everywhere

"jwt": {
  "provider": "jwt",
  "module": "passport-empty",
  "secretOrKey": "--your secret--",
  "authScheme": "saml",
  "successRedirect": "/",
  "failureRedirect": "/"
}

**Comment by DalderupMaurice** [comment]
Is this still something to be implemented/documented within composer itself?
It'd be very helpful!

**Comment by prtk418** [comment]
@derekm110 @saurabhgoel003  hey, can you help me setting maxRedirects:0 using ajax/xhr

**Comment by sunder5** [comment]
@gangachris You have stated that you have created `jwt` token outside and passing it to rest-endpoint. Could you please tell me how did create it? I mean what is the `payload` you passed to create?
```
jwt.sign({ data: payload}, SECRET, {expiresIn: config.getJwtExpiryDuration(), algorithm: 'HS256'}, (err, token) => {
            if (err) reject(err)
            else resolve(token)
        })
```
Here you have used secreate key as `gSi4WmttWuvy2ewoTGooigPwSDoxwZOy`

**Comment by djraj** [comment]
Thank you @derekm110  and @saurabhgoel003  . I finally managed to get it working in a custom Java application. I have set the redirect option to false and it works.

**Comment by oareowerei** [comment]
Has anyone managed to get this to work with Angular 2+ ? According to [this stackoverflow ](https://stackoverflow.com/questions/36885556/angular2-watch-for-302-redirect-when-fetching-resource)it can't be done @derekm110 I can see that you managed to alter the redirect behavior in your client. Might I ask what sort of framework you are using? Will I have to create some sort of extra server between the rest server and my browser client?

**Comment by mrudav-shukla** [comment]
Hi @mrosack / @gangachris , I have been able to successfully implement this JWT auth strategy using your guides. However, I have been facing some unusual issues. I have a total of 8 composer rest servers that connects to 8 different orgs on fabric. All of them has the same jwt implementation as well as their secret is also the same. However, only two of them are working as expected and others aren't working.

Basically, after exchanging the access token for the very first time using auth bearer token, a database named "test" is created in mongodb. This mongodb's database has three collections: accessToken, user and userIdentity. These collections stores the auth details for subsequent request. 

In my case, I'm able to get the access token successfully, however, no collections are created on the mongodb that can store the auth details. So what happens is, the authentication for the subsequent requests fails with the message:
> Unhandled error for request POST /api/Wallet/import: Error: Authorization Required

This is unusual because this setup is working fine on 2 out of 8 servers. I've tried restarting the entire network and also pruning the fabric network and starting it from scratch. But I'm still stuck in this. Can you please guide me out with this? What can be the reason the collections are not created on the mongodb?

**Comment by paul-thethi** [comment]
Hi @gangachris ,

Nice tutorial!

I have got the JWT strategy working. I generate JWT’s and composer authenticates the token and  return the access token.

The issue that I do have though, if you create a nested participant for example:

participant Bank identified by id {
    o String id
    o String name
}

participant BankUser identified by id { 
    o String id
    o String name
    --> Bank bank
}

When I try to generate an access token for the Bank User I get the below error:

ValidationError: The `user` instance is not valid. Details: `email` is invalid (value: “7465726d696e616c4f7065726...com”)

After doing some investigations I managed to get it to work by directly inserting a record into the userIdentity collection in mongoDb.

db.userIdentity.insert({
  "provider": "jwt",
  "authScheme": "saml",
  "externalId": "bankUser@some-network",
  "profile": {
    "timestamp": 1552475153,
    "expiry": 1553771153,
    "email": "bankUser@some-network.com",
    "name": "bankUser@oil-escrow-network",
    "id": "bankUser@some-network",
    "iat": 1552475153
  },
  "credentials": {},
  "created": ISODate("2019-03-13T11:05:52.167Z"),
  "modified": ISODate("2019-03-13T11:05:53.894Z"),
  "userId": ObjectId("5c8a48f0ac7e590012e10e6b")
});

I don’t really like the idea of having to insert directly into mongoDb for this to work.

Is there something that im missing here?
